lib/jsdom/living/nodes/Document-impl.js

```diff
@@ -56,44 +55,187 @@ const ProgressEvent = require("../generated/ProgressEvent");
 const TouchEvent = require("../generated/TouchEvent");
 const UIEvent = require("../generated/UIEvent");
 
+const { ResourceLoader } = require("../../browser/new-resource-loader");
+const { wrapCookieJarForRequest } = require("../xhr-utils");
+
 function clearChildNodes(node) {
   for (let child = domSymbolTree.firstChild(node); child; child = domSymbolTree.firstChild(node)) {
     node.removeChild(child);
   }
 }
 
+class AsyncResourceQueue {
+  constructor() {
+    this.items = new Set();
+  }
+
+  count() {
+    return this.items.size;
+  }
+
+  _notify() {
+    if (this._listener) {
+      this._listener();
+    }
+  }
+
+  setListener(listener) {
+    this._listener = listener;
+  }
+
+  push(request, onLoad, onError) {
+    const q = this;
+
+    q.items.add(request);
+
+    function check(error, data) {
+      let promise;
+
+      if (onError && error) {
+        promise = onError(error);
+      } else if (onLoad && data) {
+        promise = onLoad(data);
+      }
+
+      promise
+        .then(() => {
+          q.items.delete(request);
+
+          if (q.count() === 0) {
+            q._notify();
+          }
+        });
+    }
+
+    return request
+      .then(data => {
+        if (onLoad) {
+          return check(null, data);
+        }
+
+        q.items.delete(request);
+
+        if (q.count() === 0) {
+          q._notify();
+        }
+
+        return null;
+      })
+      .catch(err => {
+        if (onError) {
+          return check(err);
+        }
+
+        q.items.delete(request);
+
+        if (q.count() === 0) {
+          q._notify();
+        }
+
+        return null;
+      });
+  }
+}
+
 class ResourceQueue {
```

Comment:
Can we move this to its own file too?

And can we make sure the files have good explanations of each class's purpose? It's a bit confusing trying to figure out why there are both ResourceQueue and AsyncResourceQueue. Maybe they need better names?

Reply:
I will move all the classes to a new file